更换WiFi时短暂断网后不能成功重连，bot被closed

### 问题描述

网络波动后bot被closed，不知道是否与今天升级到2.14.0有关。我记得以前断网数小时，再联网后也会自动重连。现在怎么短暂断网就被closed了，想知道是新版本对重连次数加了什么上限么还是其他原因造成。

### 复现

当设备更换WiFi时可能会发生。

### mirai-core 版本

2.14.0

### bot-protocol

ANDROID_PAD

### 其他组件版本

_No response_

### 系统日志

_No response_

### 网络日志

```text
[93m2023-02-10 19:48:58 W/Net 2305339556: Exception in resumeConnection.
HeartbeatFailedException: 2305339556.AliveHeartbeat, recoverable=true, cause=PacketTimeoutException(message=2305339556.AliveHeartbeat Scheduler: Timeout receiving action response, cause=kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms)
	at net.mamoe.mirai.internal.network.handler.CommonNetworkHandler$StateOK$heartbeatJobs$1$1.invoke(CommonNetworkHandler.kt:346)
	at net.mamoe.mirai.internal.network.handler.CommonNetworkHandler$StateOK$heartbeatJobs$1$1.invoke(CommonNetworkHandler.kt:346)
	at net.mamoe.mirai.internal.network.handler.NetworkHandlerSupport.setStateImpl$mirai_core(NetworkHandlerSupport.kt:329)
	at net.mamoe.mirai.internal.network.handler.CommonNetworkHandler$StateOK$heartbeatJobs$1.invoke(CommonNetworkHandler.kt:408)
	at net.mamoe.mirai.internal.network.handler.CommonNetworkHandler$StateOK$heartbeatJobs$1.invoke(CommonNetworkHandler.kt:345)
	at net.mamoe.mirai.internal.network.components.TimeBasedHeartbeatSchedulerImpl$launchHeartbeatJobAsync$1.invokeSuspend(HeartbeatScheduler.kt:131)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:33)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:33)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
Caused by: PacketTimeoutException(message=2305339556.AliveHeartbeat Scheduler: Timeout receiving action response, cause=kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms)
	at net.mamoe.mirai.internal.network.components.TimeBasedHeartbeatSchedulerImpl$launchHeartbeatJobAsync$1.invokeSuspend(HeartbeatScheduler.kt:133)
	... 12 more
Caused by: kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms
	at kotlinx.coroutines.TimeoutKt.TimeoutCancellationException(Timeout.kt:184)
	at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:154)
	at kotlinx.coroutines.EventLoopImplBase$DelayedRunnableTask.run(EventLoop.common.kt:508)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.DefaultExecutor.run(DefaultExecutor.kt:108)
	at java.base/java.lang.Thread.run(Thread.java:833)
[0m
[0m2023-02-10 19:48:58 V/Net 2305339556: Validated caches.[0m
[0m2023-02-10 19:48:58 V/Net 2305339556: Loading server list from cache.[0m
[0m2023-02-10 19:48:58 V/Net 2305339556: Server list: 125.39.196.178:80, 59.83.208.153:80, 116.130.229.191:80, 220.194.118.217:14000, msfwifi.3g.qq.com:8080, 112.86.231.155:443, 60.29.242.161:80, 59.83.207.217:14000, 157.148.54.17:8080, 157.148.54.249:443.[0m
[92m2023-02-10 19:48:58 I/Net 2305339556: ECDH key is valid.[0m
[0m2023-02-10 19:48:58 V/Net 2305339556: Send: wtlogin.exchange_emp(10:fast-login)[0m
[0m2023-02-10 19:49:03 V/Net 2305339556: Send: wtlogin.exchange_emp(10:fast-login)[0m
[0m2023-02-10 19:49:08 V/Net 2305339556: Send: wtlogin.login(9:password-login)[0m
[0m2023-02-10 19:49:13 V/Net 2305339556: Send: wtlogin.login(9:password-login)[0m
[93m2023-02-10 19:49:18 W/Net 2305339556: Exception in resumeConnection.
kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms
	at kotlinx.coroutines.TimeoutKt.TimeoutCancellationException(Timeout.kt:184)
	at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:154)
	at kotlinx.coroutines.EventLoopImplBase$DelayedRunnableTask.run(EventLoop.common.kt:508)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.DefaultExecutor.run(DefaultExecutor.kt:108)
	at java.base/java.lang.Thread.run(Thread.java:833)
[0m
[93m2023-02-10 19:49:18 W/Net 2305339556: Network selector received exception, closing bot. (kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms)[0m
[93m2023-02-10 19:49:18 W/Net 2305339556: Exception in resumeConnection.
kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms
	at kotlinx.coroutines.TimeoutKt.TimeoutCancellationException(Timeout.kt:184)
	at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:154)
	at kotlinx.coroutines.EventLoopImplBase$DelayedRunnableTask.run(EventLoop.common.kt:508)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.DefaultExecutor.run(DefaultExecutor.kt:108)
	at java.base/java.lang.Thread.run(Thread.java:833)
	Suppressed: HeartbeatFailedException: 2305339556.AliveHeartbeat, recoverable=true, cause=PacketTimeoutException(message=2305339556.AliveHeartbeat Scheduler: Timeout receiving action response, cause=kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms)
		at net.mamoe.mirai.internal.network.handler.CommonNetworkHandler$StateOK$heartbeatJobs$1$1.invoke(CommonNetworkHandler.kt:346)
		at net.mamoe.mirai.internal.network.handler.CommonNetworkHandler$StateOK$heartbeatJobs$1$1.invoke(CommonNetworkHandler.kt:346)
		at net.mamoe.mirai.internal.network.handler.NetworkHandlerSupport.setStateImpl$mirai_core(NetworkHandlerSupport.kt:329)
		at net.mamoe.mirai.internal.network.handler.CommonNetworkHandler$StateOK$heartbeatJobs$1.invoke(CommonNetworkHandler.kt:408)
		at net.mamoe.mirai.internal.network.handler.CommonNetworkHandler$StateOK$heartbeatJobs$1.invoke(CommonNetworkHandler.kt:345)
		at net.mamoe.mirai.internal.network.components.TimeBasedHeartbeatSchedulerImpl$launchHeartbeatJobAsync$1.invokeSuspend(HeartbeatScheduler.kt:131)
		at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
		at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:33)
		at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
		at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
		at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:33)
		at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
		at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
		at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
		at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
		at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
		at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
		at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
	Caused by: PacketTimeoutException(message=2305339556.AliveHeartbeat Scheduler: Timeout receiving action response, cause=kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms)
		at net.mamoe.mirai.internal.network.components.TimeBasedHeartbeatSchedulerImpl$launchHeartbeatJobAsync$1.invokeSuspend(HeartbeatScheduler.kt:133)
		... 12 more
	Caused by: kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms
		at kotlinx.coroutines.TimeoutKt.TimeoutCancellationException(Timeout.kt:184)
		at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:154)
		at kotlinx.coroutines.EventLoopImplBase$DelayedRunnableTask.run(EventLoop.common.kt:508)
		at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
		at kotlinx.coroutines.DefaultExecutor.run(DefaultExecutor.kt:108)
		at java.base/java.lang.Thread.run(Thread.java:833)
[0m
[93m2023-02-10 19:49:18 W/Net 2305339556: Network selector received exception, closing bot. (kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms)[0m
```


### 补充信息

_No response_

--- a/mirai-core/src/commonMain/kotlin/network/handler/selector/AbstractKeepAliveNetworkHandlerSelector.kt
+++ b/mirai-core/src/commonMain/kotlin/network/handler/selector/AbstractKeepAliveNetworkHandlerSelector.kt
@@ -158,6 +158,13 @@ internal abstract class AbstractKeepAliveNetworkHandlerSelector<H : NetworkHandl
                         // 不可挽救
                         exceptionCollector.throwLast()
                     }
+                } catch (e: TimeoutCancellationException) {
+                    logIfEnabled { "... failed with TimeoutCancellationException: $e" }
+                    exceptionCollector.collect(e)
+                    close(e) // close H
+                    logIfEnabled { "CLOSED instance" }
+                    // 发包超时 可挽救
+                    return false
                 } catch (e: Throwable) {
                     // 不可挽救
                     logIfEnabled { "... failed with unexpected: $e" }

--- a/mirai-core/src/commonTest/kotlin/network/impl/common/CommonNHBotNormalLoginTest.kt
+++ b/mirai-core/src/commonTest/kotlin/network/impl/common/CommonNHBotNormalLoginTest.kt
@@ -103,7 +103,7 @@ internal class CommonNHBotNormalLoginTest : AbstractCommonNHTest() {
                 delay(1000)
             }
         }
-        assertFailsWith<TimeoutCancellationException>("test Connection reset by peer") { bot.login() }
+        assertFailsWith<TimeoutCancellationException>("test TimeoutCancellationException") { bot.login() }
         assertState(NetworkHandler.State.CLOSED)
     }
 
@@ -153,14 +153,11 @@ internal class CommonNHBotNormalLoginTest : AbstractCommonNHTest() {
                 }
             }
         }
-        val preTimes = bot.configuration.reconnectionRetryTimes
-        bot.configuration.reconnectionRetryTimes = 20
         bot.login()
         bot.network.close(HeartbeatFailedException("Heartbeat Timeout", RuntimeException("Timeout stub"), true))
         eventDispatcher.joinBroadcast()
         delay(1000L) // auto resume in BotOfflineEventMonitor
         eventDispatcher.joinBroadcast()
-        bot.configuration.reconnectionRetryTimes = preTimes
         assertState(NetworkHandler.State.OK)
     }
 

--- a/mirai-core/src/commonTest/kotlin/network/impl/common/CommonNHBotNormalLoginTest.kt
+++ b/mirai-core/src/commonTest/kotlin/network/impl/common/CommonNHBotNormalLoginTest.kt
@@ -153,7 +153,7 @@ internal class CommonNHBotNormalLoginTest : AbstractCommonNHTest() {
         bot.login()
         bot.network.close(HeartbeatFailedException("Heartbeat Timeout", RuntimeException("Timeout stub"), true))
         eventDispatcher.joinBroadcast()
-        delay(1000L) // auto resume in BotOfflineEventMonitor
+        yield() // auto resume in BotOfflineEventMonitor
         eventDispatcher.joinBroadcast()
         assertState(NetworkHandler.State.OK)
     }

--- a/mirai-core/src/commonTest/kotlin/network/impl/common/CommonNHBotNormalLoginTest.kt
+++ b/mirai-core/src/commonTest/kotlin/network/impl/common/CommonNHBotNormalLoginTest.kt
@@ -12,8 +12,10 @@
 package net.mamoe.mirai.internal.network.impl.common
 
 import io.ktor.utils.io.errors.*
+import kotlinx.coroutines.TimeoutCancellationException
 import kotlinx.coroutines.delay
 import kotlinx.coroutines.isActive
+import kotlinx.coroutines.withTimeout
 import net.mamoe.mirai.internal.BotAccount
 import net.mamoe.mirai.internal.MockConfiguration
 import net.mamoe.mirai.internal.QQAndroidBot
@@ -30,6 +32,7 @@ import net.mamoe.mirai.internal.network.handler.selector.KeepAliveNetworkHandler
 import net.mamoe.mirai.internal.network.handler.selector.NetworkChannelException
 import net.mamoe.mirai.internal.network.handler.selector.SelectorNetworkHandler
 import net.mamoe.mirai.internal.network.handler.selectorLogger
+import net.mamoe.mirai.internal.network.impl.HeartbeatFailedException
 import net.mamoe.mirai.internal.network.protocol.packet.login.StatSvc
 import net.mamoe.mirai.internal.test.runBlockingUnit
 import net.mamoe.mirai.network.CustomLoginFailedException
@@ -91,6 +94,19 @@ internal class CommonNHBotNormalLoginTest : AbstractCommonNHTest() {
         assertState(NetworkHandler.State.CLOSED)
     }
 
+    // #1963
+    @Test
+    fun `test first login failure with internally handled exceptions3`() = runBlockingUnit {
+        setSsoProcessor {
+            //Throw TimeoutCancellationException
+            withTimeout(1) {
+                delay(1000)
+            }
+        }
+        assertFailsWith<TimeoutCancellationException>("test Connection reset by peer") { bot.login() }
+        assertState(NetworkHandler.State.CLOSED)
+    }
+
     // 经过 #1963 考虑后在初次登录遇到任何错误都终止并传递异常
 //    @Test
 //    fun `test network broken`() = runBlockingUnit {
@@ -118,4 +134,34 @@ internal class CommonNHBotNormalLoginTest : AbstractCommonNHTest() {
 
         assertState(NetworkHandler.State.OK)
     }
+
+    // #2504, #2488
+    @Test
+    fun `test resume failed with TimeoutCancellationException`() = runBlockingUnit {
+        var first = true
+        var failCount = 3
+        setSsoProcessor {
+            if (first) {
+                first = false
+            } else {
+                if (failCount > 0) {
+                    failCount--
+                    //Throw TimeoutCancellationException
+                    withTimeout(1) {
+                        delay(1000)
+                    }
+                }
+            }
+        }
+        val preTimes = bot.configuration.reconnectionRetryTimes
+        bot.configuration.reconnectionRetryTimes = 20
+        bot.login()
+        bot.network.close(HeartbeatFailedException("Heartbeat Timeout", RuntimeException("Timeout stub"), true))
+        eventDispatcher.joinBroadcast()
+        delay(1000L) // auto resume in BotOfflineEventMonitor
+        eventDispatcher.joinBroadcast()
+        bot.configuration.reconnectionRetryTimes = preTimes
+        assertState(NetworkHandler.State.OK)
+    }
+
 }

--- a/mirai-core/src/commonTest/kotlin/network/impl/common/CommonNHBotNormalLoginTest.kt
+++ b/mirai-core/src/commonTest/kotlin/network/impl/common/CommonNHBotNormalLoginTest.kt
@@ -12,10 +12,7 @@
 package net.mamoe.mirai.internal.network.impl.common
 
 import io.ktor.utils.io.errors.*
-import kotlinx.coroutines.TimeoutCancellationException
-import kotlinx.coroutines.delay
-import kotlinx.coroutines.isActive
-import kotlinx.coroutines.withTimeout
+import kotlinx.coroutines.*
 import net.mamoe.mirai.internal.BotAccount
 import net.mamoe.mirai.internal.MockConfiguration
 import net.mamoe.mirai.internal.QQAndroidBot
@@ -129,7 +126,7 @@ internal class CommonNHBotNormalLoginTest : AbstractCommonNHTest() {
         bot.network.close(StatSvc.ReqMSFOffline.MsfOfflineToken(0, 0, 0))
 
         eventDispatcher.joinBroadcast()
-        delay(1000L) // auto resume in BotOfflineEventMonitor
+        yield() // auto resume in BotOfflineEventMonitor
         eventDispatcher.joinBroadcast()
 
         assertState(NetworkHandler.State.OK)

